lib/models/room.js

```diff
@@ -214,43 +292,195 @@ Room.prototype.getAvatarUrl = function(baseUrl, width, height, resizeMethod,
  };
 
 /**
- * Add some events to this room's timeline. Will fire "Room.timeline" for
- * each event added.
+ * Create a new timeline for this room
+ *
+ * @param {Array<MatrixEvent>} stateBeforeEvent state of the room before this event
+ * @return {module:models/event-timeline~EventTimeline} newly-created timeline
+ */
+Room.prototype.createTimeline = function() {
+    if (!this._timelineSupport) {
+        throw Error("timeline support is disabled. Set the 'timelineSupport'" +
+                    " parameter to true when creating MatrixClient to enable" +
+                    " it.");
+    }
+
+    var timeline = new EventTimeline(this.roomId);
+    this._timelines.push(timeline);
+    return timeline;
+};
+
+
+/**
+ * Add events to a timeline
+ *
+ * <p>Will fire "Room.timeline" for each event added.
+ *
  * @param {MatrixEvent[]} events A list of events to add.
- * @param {boolean} toStartOfTimeline True to add these events to the start
+ *
+ * @param {boolean} toStartOfTimeline   True to add these events to the start
  * (oldest) instead of the end (newest) of the timeline. If true, the oldest
  * event will be the <b>last</b> element of 'events'.
+ *
+ * @param {module:models/event-timeline~EventTimeline=} timeline   timeline to
+ *    add events to. If not given, events will be added to the live timeline
+ *
+ * @param {string=} paginationToken   token for the next batch of events
+ *
  * @fires module:client~MatrixClient#event:"Room.timeline"
+ *
  */
-Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline) {
-    var stateContext = toStartOfTimeline ? this.oldState : this.currentState;
+Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline,
+                                              timeline, paginationToken) {
+    if (!timeline) {
+        timeline = this._liveTimeline;
+    }
 
-    function checkForRedaction(redactEvent) {
-        return function(e) {
-            return e.getId() === redactEvent.event.redacts;
-        };
+    if (!toStartOfTimeline && timeline == this._liveTimeline) {
+        // special treatment for live events
+        this._addLiveEvents(events);
+        return;
+    }
+
+    var updateToken = false;
+    for (var i = 0; i < events.length; i++) {
+        var existingTimeline = this._checkExistingTimeline(events[i], timeline,
+                                                           toStartOfTimeline);
+        if (existingTimeline) {
+            // switch to the other timeline
+            timeline = existingTimeline;
```

Comment:
This is obviously the gnarly bit of this change, and you won't be surprised to learn that I ended up fiddling with it a fair bit.

I've added a wordy comment which I hope makes it much clearer wtf is going on here. In the process, I've realised that I missed an edge case, which for now I have marked with an XXX.

In terms of the overuse of `timeline` - yes I suppose it *is* being overused, but here there is only one "timeline" of interest, and that is the timeline we are currently adding events to. I think it makes sense to reuse, and clobber, the input variable here.